src/loaders/s3m_load.c

```diff
@@ -619,6 +620,11 @@ static int s3m_load(struct module_data *m, HIO_HANDLE * f, const int start)
 			xxs->flg |= XMP_SAMPLE_16BIT;
 		}
 
+		load_sample_flags = (sfh.ffi == 1) ? 0 : SAMPLE_FLAG_UNS;
+		if (sih.pack == 4) {
+			load_sample_flags = SAMPLE_FLAG_ADPCM;
```

Comment:
Any reason not to bitwise OR this?  The logic in `libxmp_load_sample` does not indicate the flags here are mutually exclusive--and if they were, this block should be refactored to prevent the redundant check/assignment.

Reply:
You're right, nice catch. The reason I implemented it like this is [that's how the copy of libmodplug 0.8.9.0 I used as a reference does it](https://github.com/AliceLR/megazeux/blob/master/contrib/libmodplug/src/load_s3m.cpp#L397), but there's no guarantee that's correct. I'll look into it more...

Reply:
Update: from what I can gather [MOD Plugin](http://efr.worldsapart.net/modreadme.html) is the source of ADPCM4 samples in S3Ms (and [OpenMPT's handling of these](https://github.com/OpenMPT/openmpt/blob/master/soundlib/S3MTools.cpp#L117) attributes them to MOD Plugin), so libmodplug's behavior is probably correct. Probably also worth noting that in the same file I linked in the previous comment, the code for saving ADPCM4 samples [similarly overrides the U8 flag](https://github.com/AliceLR/megazeux/blob/master/contrib/libmodplug/src/load_s3m.cpp#L621).

Reply:
If you still need a confirmation: Yes, ModPlug ADPCM is exclusively 8-bit mono. It must not be combined with any other flags because it is neither defined how to handle stereo ADPCM samples, nor is there any other algorithm defined apart from 4-to-8-bit expansion.